### PR TITLE
Psubscribe support

### PR DESCRIPTION
**Pull request by janoc**
Here is a basic PSUBSCRIBE/PUNSUBSCRIBE support

I have done some basic tests, seems to work, but more testing is welcome. I am also not 100% sure about the public API - I have added usePattern boolean as the last (and optional) argument to the existing subscribe/unsubscribe functions in order to avoid code duplication, but perhaps it would be better to create separate functions for it.

### CODE REVIEW

**Comment by nekipelov** [comment]
I want to add methods for all redis command later. Thats why we should keep `subscribe` semantics. Can you move `subscribe` implementation to the RedisClientImpl and add `psubscribe`?

Will be like this:

```cpp
RedisAsyncClient::subscribe(channel, messagesHandler, handler)
{
    return pimpl->subscribe("subscribe", channel, messagesHandler, handler);
}

RedisAsyncClient::psubscribe(channel, messagesHandler, handler)
{
    return pimpl->subscribe("psubscribe", channel, messagesHandler, handler);
}
```

**Comment by janoc** [comment]
OK, I have moved the code to the implementation class and added separate functions. It required a bit of because the Handle struct is defined in the RedisAsyncClient and not available in the impl class.

File: src/redisclient/impl/redisclientimpl.cpp [hunk 1]
```diff
@@ -56,12 +56,14 @@ void RedisClientImpl::doProcessMessage(RedisValue v)
     if( state == State::Subscribed )
     {
         std::vector<RedisValue> result = v.toArray();
+        auto rsize = result.size();
```
Comment (nekipelov) [inline src/redisclient/impl/redisclientimpl.cpp]: Please rename `rsize` to `resultSize`.

File: src/redisclient/impl/redisasyncclient.cpp [hunk 2]
```diff
@@ -83,123 +83,42 @@ RedisAsyncClient::Handle RedisAsyncClient::subscribe(
...
 
-        pimpl->errorHandler(ss.str());
-        return Handle();
-    }
+RedisAsyncClient::Handle RedisAsyncClient::psubscribe(
+    const std::string &pattern,
+    std::function<void(std::vector<char> msg)> msgHandler,
+    std::function<void(RedisValue)> handler)
+{
+    auto handle_id = pimpl->subscribe("psubscribe", pattern, msgHandler, handler);
```
Comment (nekipelov) [inline src/redisclient/impl/redisasyncclient.cpp]: Please rename `handle_id` to `handleId`.

File: src/redisclient/impl/redisasyncclient.cpp [hunk 3]
```diff
@@ -83,123 +83,42 @@ RedisAsyncClient::Handle RedisAsyncClient::subscribe(
...
-        return handle;
-    }
-    else
-    {
-        std::stringstream ss;
-
-        ss << "RedisAsyncClient::command called with invalid state "
-           << to_string(pimpl->state);
+    auto handle_id = pimpl->subscribe("subscribe", channel, msgHandler, handler);    
+    return { handle_id , channel };
```
Comment (nekipelov) [inline src/redisclient/impl/redisasyncclient.cpp]: Please rename `handle_id` to `handleId`.

**Comment by nekipelov** [comment]
It looks good. Can you rename some variables and remove commit e60eec030e1d6d28c3e88d11eef640f4066894fb (rebase and squash)?

**Comment by janoc** [comment]
I have renamed the variables as you wanted. However, I didn't remove that commit, because the branch would stop building. 

Either cherrypick the commits you want or we need to first fix and merge #32 and then rebase this on it. It will save me the work of having to fix the build twice.

**Comment by janoc** [comment]
Nope, it doesn't build, because of the duplicate symbols found (the autolinking issue). Will do a temporary hack on my side that will not be included in the pull req. to get it to build until #32 is merged.

**Comment by janoc** [comment]
OK, done - AppVeyor CI may not build because of the double linking in Windows, that is to be expected.

**Comment by nekipelov** [comment]
Excellent. Thank you!

**Pull request merged**